### ISSUE TITLE
Compile time, created files size, and exec time all seem large (Win10/mingw64 in conda)

### ISSUE THREAD

**Issue by DanCodigaMWRA** [Nuitka/Nuitka#599]
My end goal is to be able to use nuitka to create standalone executables from python scripts that import and use several common packages (numpy, pandas, xarray, and matplotlib.pyplot). 

I am using nuitka v. 0.6.5 on Windows 10 (x86_64) with Miniconda and python 3.6.7. I installed the C compiler using conda forge and m2w64-gcc package. I'm doing everything within a conda env, in which I have the various packages installed (numpy, pandas, xarray, matplotlib, etc) in addition to nuitka and mingw64.

I did some tests on simple hello.py type examples and the results seem to have very long compile times, very large sizes of resulting files, and long execution times for resulting files. 

I'm trying to understand if there is something sub-optimal about my environment/configuration and/or the way I am calling nuitka-- as opposed to these performance concerns just being typical for nuitka, given the challenge to include and use all these large packages at once.

I set up a series progressively more demanding examples:
* hello1.py is the one in the tutorial (no imports, very simple)
* hello2.py imports numpy and executes one or two simple calls from it
* hello3.py also imports pandas and executes one or two simple calls from it too
* hello4.py also imports xarray and executes one or two simple calls from it too
* hello5.py also imports matplotlib.pyplot and executes one or two simple calls from it too

Note that hello5.py is an example of my end goal, as noted above. Maybe this end goal is simply not realistic for nuitka? Is anybody using nuitka for comparable purposes, without trouble?

An example of the compile command (for hello3.py) is as follows:

"python -m nuitka --standalone --mingw64 --follow-imports --plugin-enable=tk-inter --plugin-enable=qt-plugins --plugin-enable=pylint-warnings --plugin-enable=torch --plugin-enable=numpy hello3.py"

There are some warnings (no version of Visual Studio compiler...) during compilation, but at least for the first two examples, they are benign.

For the first test, hello1.py (no imports) the results are:
* compile time = about 10 seconds
* size of the resulting files (combined "build" and "dist" folders)  = about 25 megs
* execution time of the resulting executable =  trivial (as when executing uncompiled code)

For the second test, hello2.py (import and use numpy) the results are:
* compile time = several minutes
* size of the resulting files (combined "build" and "dist" folders)  = about 744 megs
* execution time of the resulting executable =  many seconds (far longer than when executing uncompiled code)

For the third test, the compilation failed with the error message

OSError: [WinError 145] The directory is not empty: 'hello3.dist\\PyQt5\\qt-plugins\\scenegraph'

but that was after 10s of minutes of compile time (!), and after the build/dist folders were already about 926 megs (!). 

Unless there is something fundamentally wrong with my configuration and/or the syntax of my compilation command, I am led to believe that nuitka won't be able to help me reach my goal, because the compile times, storage size of resulting files, and executable file execution time are all so large as to be impractical. Correct?

Other ideas for how to approach my end goal are welcome, if it is indeed feasible with nuitka. If I am missing something obvious it would be great to learn what it is, and take a different tack.

**Comment by JorjMcKie** [comment]
Please be aware that those imported packages are among the largest you can get: **numpy** - 50 MB, **scipy** - 100 MB, **pandas** - 44 MB, **matplotlib** - 22 MB, **PyQt** - 233 MB, **torch** - 1.3 GB (!!!).

A compiler like Nuitka, parsing the source code, **_cannot know_** which part(s) of all those packages (and the ones imported some levels deeper) will actually be used during runtime.
If allow me a kidding comment: the set of plugins you added yourself also is on the "better safe than sorry" side.

To inject runtime information that helps cutting down the to-be-compiled volumes, we have developed a utility, called "hinted compilation". Please see [here](https://github.com/Nuitka/NUITKA-Utilities/tree/master/hinted-compilation). There is also some theory background on how this works.

Just 2 days ago I compiled a simple matplotlib app using several backend options (Qt, tkinter, WxPython).
Some idicative results (Windows, using tkinter as backend)

| method | time | dist size |
| -------- | ----- | ------- |
| vanilla | 50 min | 250 MB |
| hinted | 16 min | 120 MB |

So compile times and output size are much smaller, if Nuitka is being given the chance to confine to the actually used parts: If scipy is not used, it won't get into the dist folder.
Another advantage of hinted compilation is, that it detects the required plugins and generates the corresponding parameters.
Takes out a lot of hassle and shortens the command line.

My recommendation therefore is to also follow this path.

**Comment by DanCodigaMWRA** [comment]
Jorj thank you for the quick reply and explanation. I updated to nuitka v 0.6.6 and used hinted compilation following the instructions at the link you sent.  It worked fine (except for my tests with matplotlib, see at bottom below) and the results, shown below, are improved.

I’ll have to do more testing to see if the results are similar when the script is “real” (in the sense that it does more than a couple of quick calls, to basic functionality for each package, as in the tests I ran).

But (assuming that makes relatively minor changes) in terms of the end goal I’m working towards it seems like nuitka may be able to get me there-- as long as I am willing to accept that the executable files occupy up to roughly a gigabyte or more space on disk, and that execution times for the compiled code are considerably slower than uncompiled. If you think I should be doing better, I welcome further suggestions from you.

(That kind of performance is better then what I found for pyinstaller and cx_freeze-- both of which I had been working with earlier, but found to be very fussy and hard to use, and also to create huge files after extremely long compile times, which oftentimes did not execute without errors. )

Here are the results:

1. hello1.py (no imports)
- compile time = under a minute
- size on disk (hello1.dist folder) = 24 megs
- execution time = a couple seconds

2. hello2.py (imports/uses numpy only)
- compile time = a few minutes
- size on disk = 625 megs
- execution time = several seconds

3. hello3.py (imports/uses numpy & pandas)
- compile time = 10s of minutes
- size on disk = 720 megs
- execution time = several seconds

4. hello4.py (imports/uses numpy, pandas, & xarray)
- compile time = 10s of minutes
- size on disk = 730 megs
- execution time = 10s of seconds 

I did not succeed in any cases that included matplotlib. Even for this simple matplotlib-only test case,

def talk(message):
    return "Talk " + message
def main():
    print( talk("Hello World"))
    import matplotlib.pyplot as plt
    plt.plot([1,2,3],[3,2,1])
    plt.savefig('hello6.png')

although get-hints.py and nuitka-hints.py compilation would succeed, during execution there was this error: 

ImportError: cannot import name 'pyplot'

You mentioned that you recently did a test with a matplotlib app, so I shouldn’t be getting such an error should I? Could you point me to examples that include matplotlib successfully, or any tips for doing so? 

p.s. Your "kidding comments" are most welcome, I deserve them! I knew the list of plugins I had included in my earlier command line compile statement was "everything including the kitchen sink". I did that simply because I don't understand much of the details, in terms of which plugins to include or exclude, etc—I’m not even using torch. So I was indeed just trying to be safe rather than sorry.

**Comment by JorjMcKie** [comment]
> You mentioned that you recently did a test with a matplotlib app, so I shouldn’t be getting such an error should I? Could you point me to examples that include matplotlib successfully, or any tips for doing so?

Yes, for the moment, you should expect a solution to your matplotlib issues.
Depending on how inclined you are to experimenting a little more, I can send you the required updates, so that you do not need to wait for the official release cycle.

Some background for those issues:
* Matplotlib supports almost 20 different backends as output. The default backend and a few dozens of other parameters for matplotlib are contained in config file ``matplotlibrc``, which is located in ``matplotlib/mpl-data`` folder of the installation.
* Any installation may optionally entertain its own ``matplotlibrc`` located at some other place.
* If a statement like ``backend: <string>`` is missing in that file, matplotlib tries half a dozen of alternatives, which depend on operating system and installed packages (PyQt, WxPython, MacOSX, Tkinter, ...). If you want a specific one you can specify it via ``matplotlib.use(<string>)``. But this must **_happen before_** ``pyplot`` is imported!
* Nuitka has issues following the consequential hidden imports (based on both, implicit or explicit) backend selection. We could of course do implicit imports for simply all of those 20 - which we don't want to do for space implications. One must be aware, that any specific import of a backend entails a handful of yet other imports ... Here, hinted compilation comes in handy, as it logs which backends are actually selected.

**Comment by JorjMcKie** [comment]
Another comment on your final ``dist`` sizes: they seem high to me even after hinting.
Presumably you are using Intel's MKL, don't you? They alone are half a GB in size.

As an occasional user of numpy / scipy curve fitting solutions, I always found the basic OpenBlas 
components sufficient coming with numpy and scipy directly from PyPI. numpy from there only has one or two shared libraries (1 on Windows 32 MB, 2 on Linux totalling 30 MB).

I never saw any significant performance differences to using MKL ... admittedly I am not doing professional work in that area though.

**Comment by JorjMcKie** [comment]
> ... the compiled code are considerably slower than uncompiled.

**That** should actually not happen! And derserves investigation.
It is ok to not see significant improvements, because the science packages you are importing, are already written in C for the most part, and wrapped with a thing Python layer.

**Comment by DanCodigaMWRA** [comment]
Super helpful info thank you.

> Yes, for the moment, you should expect a solution to your matplotlib issues.
> Depending on how inclined you are to experimenting a little more, I can send you the required updates, so that you do not need to wait for the official release cycle.

This depends on how long you expect for it to be, before the official release incorporates the updates. If it's a month (or less) away, I'll wait. Glad to know it's something that can be overcome.

> Another comment on your final `dist` sizes: they seem high to me even after hinting.
> Presumably you are using Intel's MKL, don't you? They alone are half a GB in size.
> As an occasional user of numpy / scipy curve fitting solutions, I always found the basic OpenBlas
> components sufficient coming with numpy and scipy directly from PyPI. numpy from there only has one or two shared libraries (1 on Windows 32 MB, 2 on Linux totalling 30 MB).

I have a limited knowledge of this. I believe that yes I am using MKL rather than OpenBLAS.

I am working within an Anaconda environment (on Win10 x64) built using conda-forge as the primary channel (I don't get things directly from PyPI; or, if conda does that for me, I view it as "under the hood"). I doubt I would notice a performance difference between MKL and OpenBLAS-- for me the priority is the convenience of generating the environment using conda. Here is the yml file I used with conda, to generate the environment I'm using (called "Nuitka") to do all the tests I have been describing:

name: Nuitka
channels:
  - conda-forge
  - defaults
dependencies:
  - python=3.6
  - libpython
  - m2w64-gcc
  - matplotlib
  - nuitka
  - numpy
  - pandas
  - xarray

If you want to suggest how I could set up the yml file differently, such that I end up with OpenBLAS and/or other aspects that could improve performance of nuitka in the tests I'm running, please do. Again, I am not particularly in need of MKL vs OpenBLAS-- my main priority is just to be able to generate the environment conveniently using conda and a yml file.

> > ... the compiled code are considerably slower than uncompiled.
> 
> **That** should actually not happen! And derserves investigation.
> It is ok to not see significant improvements, because the science packages you are importing, are already written in C for the most part, and wrapped with a thing Python layer.

Here is what I find, using example hello4.py. First I run it uncompiled, "python hello4.py". It takes at most a couple of seconds to complete. The first step is printing "Talk Hello World" to the screen. 

Next, I cd into the hello4.dist folder and run "hello4.exe". It takes about 10-15 seconds for "Talk Hello World" to appear on the screen, and then it takes about 10-15 more seconds to finish.

**Comment by JorjMcKie** [comment]
You very probably have those MKL libraries. I have a (mini) conda installation myself, where they can be found at ``<conda dir>\Library\bin``. They all start with ``mkl_*``, their total size is 534 MB.
They will be copied to the ``dist`` directory; no difference between Nuitka or any other tool like PyInstaller here.
I am not sure how to slim this down. A terrible waste of storage, that is for sure. I would try this:
* uninstall numpy and scipy: both are interconnected as far as these libraries are concerned
* then re-install both, but **not via conda:** I would use pip for this. I seem to remember some chat in Stackoverflow, where hints were given on how to avoid MKL. Maybe the Anaconda website has some help on this, too. If someone is not really doing serious science work, those highly optimized DLLs are not necessary.

Can I have your hello4 script please? I would like to repeat your speedtest with my vanilla numpy installation.

**Comment by JorjMcKie** [comment]
On a parallel chat in the utilities repo with @deajan we are currently discussing option to speed up compilation. Maybe you want to have a look or even join in: https://github.com/Nuitka/NUITKA-Utilities/issues/9.

Once you have broader experience with Nuitka, you will probably have modules which already are in some stable state.
Once this is the case, I would recommend compiling those as separate binary Python extensions (Python DLL having extension .pyd) using Nuitka.
This will **_significantly_** speed up the compile of main applications which use those modules ...

Giant packages like torch and tensorflow have a plethora of submodules and follow the strategy importing them all during their startup - I once counted 60+ import statements in tensorflow.
This will of course lead to giant standalone distribution sizes and extended startup times - no matter whether you use PyInstaller or a real compiler like Nuitka.

While writing this, a possible explanation for your runtime elongation came in my mind:
Might be that the compiled code loads the MKL binaries at startup already - which does not happen in interpreted mode ... I will find out and discuss with Kay.

**Comment by DanCodigaMWRA** [comment]
Yes, I don't doubt that I have MKL. (And btw I'm also using miniconda...) 

My hello4.py script:
```
def talk(message):
    return "Talk " + message

def main():
    print( talk("Hello World"))
    from numpy import ones, shape
    print('size = %s' %str(shape(ones([20,3]))))
    import pandas as pd
    df = pd.DataFrame([1,2,3,4])
    print(df)
    import xarray as xr
    ds = xr.Dataset(df)
    print(ds)
    input('hit enter when done')
```
The only reason I'm aware of what MKL/OpenBLAS are is that about a year ago when using conda there were some problems with it trying to substitute one for the other and finding incompatibilities etc. Never quite understood it but thankfully other folks figured out the trouble and fixed it; however, I believe it was partly why I now use conda-forge as my primary channel. 

As I mentioned, those kinds of things (MKL, OpenBLAS, conda-forge) are all or mostly "under the hood" for me. I try to keep it that way because I'm a scientific programmer not a true developer/coder. (And though I'm a scientist, I would be very surprised if any of my calcs are sensitive to differences between MKL and OpenBLAS...) I write scripts that use packages like numpy, pandas, xarray, matplotlib and my end products are new calculations and resulting plots (as opposed to new applications, functions, classes, etc). I do most everything in a single conda env (where I install all sorts of packages together, therefore my heavy reliance on conda) using the Spyder IDE. The reason I'm interested in nuitka is so that I might someday be able to conveniently give colleagues executable files they can run to do calcs and make plots without needing to know what python or Spyder are. 

So in this context your suggestions to use pip instead of conda, and to compile separate binary extensions etc, are not very attractive and I'm unlikely to go down that path. I guess if the code I want to compile is simple enough to only rely on a couple packages, using pip might be worth a try but because that's uncommon I think I had better stay with conda. And in terms of binary extensions I've never done that sort of thing so unless it is pretty basic (as in, not much more complicated than the steps you sent, that I followed in order to start using "hinting") I'm unlikely to dive in. It really depends on whether the goal to share executables with others reaches a mission critical status here for me.

**Comment by JorjMcKie** [comment]
I understand.
If I come across information on how to avoid MKL while still using conda, I will let you certainly know.

Currently testing `hello4.py` and found some weird things (outside the script itself):
* the script does not use `matplotlib` at all, and also does not output anything other than to `sys.stdout`, **but still** matplotlib is included by hinted compilation! Guilty are packages `pandas` and `xarray`, which import matplotlib ... however only to check if it is there - and my poor bastard **_hinted compilation_** has no choice but to obey 😟🤔
* Package `xarray `obviously is dependent on some internal data files contained in its folder ``static``. My compiled hello4.exe gets an exception because it does not find one of those files. Actually, you should never have been able to execute your hello4 ?!?

Anyway: these are interesting findings, which must betaken care of.

**Comment by JorjMcKie** [comment]
Here are first results:

* was an easy change to get hello4.py running
* the `dist` folder size currentl is 168 MB using a **non-MKL numpy** Python 3.7 64bit Windows 10, compile time 14 minutes (8 core 2.6 GHz Intel, 16MB)

This is your script with some builtin timing:

```python
def talk(message):
    return "Talk " + message


def main():
    print(talk("Hello World"))
    from numpy import ones, shape

    print("size = %s" % str(shape(ones([20, 3]))))
    import pandas as pd

    df = pd.DataFrame([1, 2, 3, 4])
    print(df)
    import xarray as xr

    ds = xr.Dataset(df)
    print(ds)


import time

t0 = time.time()
main()
t1 = time.time()
print("Duration: %g seconds" % round(t1 - t0, 3))
```

* timings: made 3 runs each. The compiled version is 3 times faster after the first run, whereas the interpreted version stays the same (all numbers are seconds).

| interpreted | compiled |
| ----------- | --------- |
| 1.891 | 2.02 |
| 1.857 | 0.607 |
| 1.887 | 0.611 |


* I will now repeat all that with the conda environment.
* As a last action, I will try to have hinted compilation kick out matplotlib. I am optimistic with this, because for a serious usage of matplotlib you would **_always_** use some backend, wouldn't you? The hinter logfile however betrays, that no matplotlib backend is ever used by your script. I intend to exploit this fact and turn down matplotlib inclusion in such circumstances.

I will keep you posted.

**Comment by DanCodigaMWRA** [comment]
Wow it gets richer and richer doesn't it! What you're learning with these tests is very interesting indeed. Silly me I had thought that my overall goal here (apply nuitka to scripts that use all these common packages together) would be such an ordinary garden variety thing to do, that you folks would already have tried to work on my use case a fair amount. Hopefully what you are learning is useful to nuitka devs and end users in addition to helping me. (I guess there aren't too many scientific programmer types like me who want to be able to create executables from these kinds of python scripts. Either that or there is a better-known way to do it, which everybody else is already using and I'm the only one in the dark...)

I was aware that pandas and xarray both build in a lot of numpy under their hoods, so to speak (though I understand almost nothing of how this actually happens). Similarly xarray also builds in a lot of pandas under its hood, and I think pandas and xarray also build in a lot of matplotlib under their hoods somehow. 

By the way, hello4.py stops short of using matplotlib and hello5.py is designed to do all of what hello4.py does but then add a simple matplotlib call at the end. In case helpful, here is hello5.py:

```
def talk(message):
    return "Talk " + message

def main():
    print( talk("Hello World"))
    from numpy import ones, shape
    print('size = %s' %str(shape(ones([20,3]))))
    import pandas as pd
    df = pd.DataFrame([1,2,3,4])
    print(df)
    import xarray as xr
    ds = xr.Dataset(df)
    print(ds)
    import matplotlib.pyplot as plt
    plt.plot([1,2,3],[3,2,1])
    plt.savefig('hello5.png')
    input('hit enter when done')
```
As I noted in my earlier post above, I never succeeded in getting the hello5.exe to run, the error related to not finding pyplot.

As far as a backend for matplotlib, I'm not even sure what gets used on my system, so I'd have to look in to that a bit. 

Finally, I poked really quickly around w/r/t forcing non-MKL to be used and that may not be too hard to accomplish within conda after all.

Thank you for reporting out what you've learned so far and continuing to dig deeper on these topics.

**Comment by JorjMcKie** [comment]
Ok, everything is fine then.

As I have been struggling with proper matplotlib support just over the recent days I can tell, that hello5 will of course (and rightfully so) cause matplotlib to be included ... **_and_**:
* either trigger **PyQt5** to also be included in a **conda** environment,
* or trigger **Tkinter** to also be included in a **non-conda** environment.

To me, this was an amazing finding over these last days.

But as I said. Let me finish the comparison with the MKL environment and also make sure, that a "fake" mentioning of matplotlib is detected and ignored.

**Comment by JorjMcKie** [comment]
> As far as a backend for matplotlib, I'm not even sure what gets used on my system, so I'd have to look in to that a bit.

I can answer this:
If you ever use ``pyplot``, then matplotlib will use the default backend. In a conda ambiente it will scan through a list of about 6 alternatives. As your system has PyQt installed and because this option is near the top of that list, it will be selected.
You also have the option to explicitely pick one via the function ``matplotlib.use(<string>)``.
Where "<string>" may be one of "Qt5Agg", "WxAgg", "TkAgg", etc. (PyQt5, wxPython, Tkinter).
If you choose to do that, it must happen before you import ``matplotlib.pyplot``.

In a non-conda environment, Tkinter ("TkAgg" string) will be selected automatically upon import of pyplot.

**Comment by JorjMcKie** [comment]
Here is the comparison for hello4.py in a conda environment:

* ``dist`` size 737 MB, compile time 46 minutes
* timings

| interpreted | compiled |
| --- | --- |
| 1.432 | 2.539 |
| 1.398 | 0.669 |
| 1.395 | 0.69 |

So you see, that I cannot confirm your alarming observation of elongated runtimes for compiled stuff. My result points in the opposite direction: after some housekeeping effort at startup (probably loading all those shared libraries), compiled code should outplay its advantages and asymptotically be twice as fast.

Interesting also, that this compiled version seems to be slower than the non-conda case.

I don't know what your assessment is regarding disk space requirement, but for me a 170 MB vs. 740 MB in the conda case is a very strong argument to either move away from conda or re-install numpy and scipy (both must be done in sync) without MKL by all means.

My next post will be about getting matplotlib out of the game.
hello5 will follow yet one step later.

**Comment by JorjMcKie** [comment]
Here are the results for hello4.py without matplotlib:

1. Non conda
    - size: 133 MB (4 MB less), copile time about the same
    - runtime average 0.44 seconds (25% faster)

2. Conda
    - size: 704 MB (33 MB smaller) compile time 37 minutes (9 minutes shorter)
    - runtime average 0.52 seconds (25% faster)

**Comment by JorjMcKie** [comment]
# Overview and Comparison for `hello5.py`
# Non Conda

* Compile time 28 minutes
* Folder size 177 MB
* Runtimes:

| Interpreted | Compiled |
| --- | --- |
| 1.628 | 3.045 |
| 1.533 | 1.032 |
| 1.543 | 0.788 |

# Conda

* Compile time 45 minutes
* Folder size 807 MB
* Runtimes:

| Interpreted | Compiled |
| --- | --- |
| 4.269 | 3.564 |
| 1.268 | 0.876 |
| 1.335 | 0.844 |

# Comments
* In this case, the two environment (conda vs. non.conda) runtimes cannot be compared directly, because under Windows the default backend Tkinter is pre-selected in `matplotlib`, whereas in conda, matplotlib will scan through several alternatives to pick the first supported backend. In the current setup, these were Tkinter for non-conda and PyQt5 for conda.

* A similar trend can be observed as with ``hello4.py``: the test runs 2 and later of the compiled version seem to be twice as fast as the interpreted version.

**Comment by DanCodigaMWRA** [comment]
Okay thanks for doing the testing, and reporting these results. All quite interesting, especially that the executable runs faster the 2nd/3rd time!
 
I was able to look at this again today. With some work, I was able to create a conda env that has no MKL in it. I re-ran the series of tests. I found the size of the dist folder is now basically on par with the ones that you have reported (differences likely because I’m using MINGW not MSVC, but not very large). So that indicates I can still be in a conda env, as long as it doesn’t have MKL, and get similar dist folder sizes as you do.

(About MINGW vs MSVC, I like MINGW because it is so simple to just include m2w64-gcc when I create my conda env. Do you know if I can do something similar with MSCV? Not very knowledgeable about these choices...)

Also the compile times I’ve been getting are roughly on par with the ones you reported, actually shorter in some cases.

However my execution times were still very long. But I think I figured out why: so far, all my work has been on a shared drive. 

I put the test files on to my local hard drive instead, and ran the tests again. As far as I can tell, this brought my execution times way down, to basically be in line with yours. Makes sense I guess—the slowness was all the remote disk reading/writing. Also I confirmed that the 2nd/3rd time I run it, the execution is much faster, like you've found.

So, all in all, I’m feeling like I understand what’s going on a lot better, and should likely be able to use nuitka for my purposes. 

I suggest that early in the documentation for Nuitka, you consider adding some statements along the following lines, as long as they aren't inaccurate (this kind of info may be obvious to you, but would have helped me so much, if I had it when I began trying to use Nuitka):

"For nuitka when using larger packages (like numpy, pandas, xarray, matplotlib) on a typical Windows PC (such as 3Ghz, 6 cores, 16 megs RAM): 
- To keep compile times, and size on disk of compiled files, reasonable:
  - Be sure your code is not relying on MKL-linked versions of the packages.
  - Use hinting.
- Compile times of order 10s of minutes, and compiled files as large as a few 100s of megabytes, are typical.
  - Without hinting, and with MKL-linked versions: compile times can be of order an hour, and compiled file sizes can be of order a gigabyte. 
- To avoid slow execution time of compiled code, keep all your files on your local hard drive (not a shared drive).
- Be aware that when you run the compiled code, the execution times will generally become significantly shorter after the first and second time."

Now some remaining questions I still need to get past, before I have a working hello5.py:

1. For hello4.py I had no errors during compilation, but I am getting a runtime error as follows, at the call to import xarray—any suggestions?

  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 656, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 626, in _load_backward_compatible

2. For hello5.py I am still getting the compile-time errors. Currently "OSError: [WinError 110] The system cannot open the device or file specified." To get past these I will need your latest matplotlib-ready version, correct? I guess that is not included in the latest 1/6/2019 version of nuitka (0.6.6) that I am using, right? If not, by what date do you estimate they will be included in a stable release?

**Comment by JorjMcKie** [comment]
Congratulations - you have made a lot of progress indeed, and I am glad you find Nuitka so useful!

I agree to your conclusions, just a few more comments:
1. Compiler: I am certainly no expert here in contrast to Kay. I am using MSVC primarily because I am a Microsoft guy in general. The second reason is that for Nuitka as a project, running a second compiler (everyone else is using mingw) creates a larger test coverage. I also have the impression, that MSVC creates smaller code than mingw albeit compile times are somewhat longer.
So I think you should be fine with what you have.

2. Compile times: we are putting constant focus on improving these. Just recently someone discovered, that the second compilation of a script can about two times or more faster when **not removing** the ``build`` folder. I have made this change in script ``nuitka-hints.py``.

3. Execution times: with embarassment I have to admit that I should have used ``time.perf_counter()`` instead of ``time.time()`` to measure the performance. The latter is just the wall clock, whereas ``perf_counter()`` collects the time attributable to the application. This function is also used by the ``%timeit`` command in IPython.
And of course: the faster the storage device of the ``dist`` folder, the shorter the exec time.

4. Your remaining errors around `xarray` and `matplotlib`: yes, I am pretty sure they will go away once you have applied my changes to your nuitka environment. If you want I can send them to you ...

5. When packaging your application(s) for distribution, a few recommendations.
    - Use a compression tool to compact the `dist` folder. I found 7zip being the best. Like all the alternatives it also lets you create self extracting archives.
    - You can merge the ``dist`` folders of several separate applications into one. Given you are doing scientific work, your applications should more or less all be using pretty much the same scientific Python packages. A merged ``dist`` folder would therefore contain all these just once obviously, but multiple ``exe`` files - one per each application.
    - There are more tools in the "Nuitka-Utilities" repo which help you doing all this stuff.

A great thank you for your documentation suggestions. I will find a prominent place to include them.

**Comment by DanCodigaMWRA** [comment]
Ok good thanks again Jorj. 

The info in your #5 could be especially useful. I could maybe have a single large dist folder with several executables that all use it-- instead of a large dist folder for every executable I want to share. I had no idea that might be possible, but if so it could be a huge improvement. 
 
About your #4. Please do send along those instructions and/or files-- unless the timeframe for them being part of an upcoming stable release is within a couple weeks of now.

**Comment by JorjMcKie** [comment]
Re #4:

Here is a copy of all standard plugins:
[standard.zip](https://github.com/Nuitka/Nuitka/files/4060014/standard.zip)
Not all of them will be different from the ones you have, but simply unzip and copy-replace them all to the following location: ``...\Lib\site-packages\nuitka\plugins\standard``.
And here are again the scripts for hinted compilation - they also are affected by the above changes:
[hinting.zip](https://github.com/Nuitka/Nuitka/files/4060029/hinting.zip)

------------------------------

Re #5: 

There is a [script](https://github.com/Nuitka/NUITKA-Utilities/blob/master/exe-merger.py), which merges one dist folder ("input") into another ("output"). It uses the awesome GUI package ``PySimpleGUI``:
![grafik](https://user-images.githubusercontent.com/8290722/72359865-735aa900-36c5-11ea-998b-824e8a97e3b2.png)

A number of compatibility checks are done to make sure both are based on the same Python- and bit-level, etc.
A number of compatibility checks are done to make sure both are based on the same Python- and bit-level, etc.
If you are sure what you did and if there are incompatibility complaints, check the "force" box. Typical false compatibility alarms occur, when different system DLL versions are chosen by the different Nuitka compile runs. Numpy, matplotlib and others are packaged with Windows common runtime DLLs which may not be identical.

-------------------------------------

Yet another [script](https://github.com/Nuitka/NUITKA-Utilities/blob/master/link-maker.py) is good for scanning through some ``dist`` folder (which may also be a merger of several, see before), finding application ``.exe`` files and putting links of those on the Windows desktop. This one only works on Windows.

Example:
My hello4 ``dist`` is 173 MB, my hello5 177 MB, which have a total size of 350 MB. Merging ``hello4.dist`` into ``hello5.dist`` gives me 237 MB (delta 137 MB).


It is always safe (and without performance penalty) to pack ``.exe`` files with the UPX utility:

```
dir *.exe
...
09.01.2020  06:25        62.733.824 hello4.exe
09.01.2020  07:01        89.463.296 hello5.exe
               2 Datei(en),    152.197.120 Bytes
...

upx -9 *.exe
                       Ultimate Packer for eXecutables
                          Copyright (C) 1996 - 2018
UPX 3.95w       Markus Oberhumer, Laszlo Molnar & John Reiser   Aug 26th 2018

     File size         Ratio      Format      Name
--------------------   ------   -----------   -----------
62733824 ->  16676864   26.58%    win64/pe     hello4.exe
89463296 ->  23566336   26.34%    win64/pe     hello5.exe
--------------------   -------  -----------   -----------
152197120 -> 40243200  26.44%                 [ 2 files ]

Packed 2 files.

...
09.01.2020  06:25        16.676.864 hello4.exe
09.01.2020  07:01        23.566.336 hello5.exe
               2 Datei(en),     40.243.200 Bytes
```
This brings down the combined dist folder size to 130 MB ... a very agreeable size!

**Comment by DanCodigaMWRA** [comment]
Okay I got those new files you provided, and started using them thank you.
I used them on a new series of the same tests as earlier. There are unfortunately still problems to overcome, with xarray and matplotlib. Both are run-time errors, i.e. compilation completes successfully.

**Xarray**

The xarray problem is what I described above (for hello4.py). So to isolate xarray I created hello7.py (hello6.py is described below) as a simpler version of hello4.py which contains only the xarray part (not numpy nor pandas).

hello7.py:

```
def talk(message):
    return "Talk " + message
def main():
    print( talk("Hello World"))
    import xarray as xr
    ds = xr.DataArray([1,2,3,4,5])
    print(ds)
    input('hit enter when done')
```

It compiles seemingly fine but during execution it gives the same error I noted above:

```
File "", line 971, in _find_and_load
File "", line 955, in _find_and_load_unlocked
File "", line 656, in _load_unlocked
File "", line 626, in _load_backward_compatible
```

**Matplotlib**

To isolate this I created hello6.py, with matplotlib only (no numpy/pandas/xarray):
```
def talk(message):
    return "Talk " + message
def main():
    print( talk("Hello World"))
    import matplotlib.pyplot as plt
    plt.plot([1,2,3],[3,2,1])
    plt.savefig('hello6.png')
    input('hit enter when done')
```
There is no compile-time error but during execution it throws this error:

```
Traceback (most recent call last):
  File "C:\Users\Codiga_D\CodigaLocalStorage\nuitka\hello6.dist\hello6.py", line 21, in <module>
  File "C:\Users\Codiga_D\CodigaLocalStorage\nuitka\hello6.dist\hello6.py", line 14, in main
  File "C:\Users\Codiga_D\CodigaLocalStorage\nuitka\hello6.dist\matplotlib\pyplot.py", line 2349, in <module matplotlib.pyplot>
  File "C:\Users\Codiga_D\CodigaLocalStorage\nuitka\hello6.dist\matplotlib\__init__.py", line 832, in __getitem__
ImportError: cannot import name 'pyplot'

```
What do you recommend that I do, in order to troubleshoot these and get past them? 

You don't have these issues, on Win10 inside a conda environment, correct? 

It could be differences in our conda environments. I provided above the env.yml file that I used to create the conda env I am in. How does yours differ from that? I did also have to tinker with it to get rid of MKL though. I believe it was

```
conda remove mkl
conda install "blas=*=openblas" -c conda-forge
```

Otherwise I guess it could be that I'm using MINGW and you're on MSVC?

**Comment by JorjMcKie** [comment]
Both, the xarray and the matplotlib issue look like you did not install (overwrite) the plugins provided via ``standard.zip``, at least not in the right Python.
E.g. when you navigate to ``nuitka/plugins/standard`` (in the Python in use for the compile ... 🤔) and open file ``DataFileCollectorPlugin.py``, then look at this position:

![grafik](https://user-images.githubusercontent.com/8290722/72385659-5d191100-36f6-11ea-9519-69f7f3ea36d7.png)

If you don't see the xarray entry, then the modification went wrong. The entry takes care of copying data files for xarray. Your error indicates that they are not there. If if had worked, then the ``dist`` folder layout should like like:

```
hello4.dist
    lxml
    numpy
    pandas
    pytz
    xarray
        static
            css
            html
```


Similarly for file ``NumpyPlugin.py``. This should look like

![grafik](https://user-images.githubusercontent.com/8290722/72385899-ca2ca680-36f6-11ea-9639-dd5a465220b9.png)

The error you are getting here means that the file ``matplotlibrc`` found in ``dist`` folder position ``hello5.dist\matplotlib\mpl-data\matplotlibrc`` has not been created by the numpy plugin. If it were so, you would find this content in that file:
![grafik](https://user-images.githubusercontent.com/8290722/72386810-74f19480-36f8-11ea-8abc-b247a65fa5bd.png)

**Comment by JorjMcKie** [comment]
> which contains only the xarray part (not numpy nor pandas)

xarray imports both - so you will end up with a dist folder having them too.

**Comment by DanCodigaMWRA** [comment]
Doh! Apologies. Indeed I overwrote those plugin files in the folder under the wrong conda env. 

Glad you could see my mistake-- and that it was such an obvious fix.

Will let you know what happens with a new series of tests as soon as I can...

**Comment by DanCodigaMWRA** [comment]
Okay so I'm in quite good shape now in terms of the overall goal I started out this issue with. 
 - I've got a compiled hello5.py (using simple calls to each of numpy, pandas, xarray, and matplotlib) and it executes quickly.
 - the compile time was around 10-15 minutes
 - the dist folder size is 296 megs

Very much appreciate all the help Jorj. Next steps, I will try things like UDX, merging in to a single dist folder, etc. But for now I can close this issue.
